File: core/nodesDev/gfToolsNodes/src/n_gfDebugPv.cpp
#include "../headers/n_gfDebugPv.h"

#define INPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(true);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(true);	\
	FNATTR.setKeyable(true);	\

#define OUTPUT_ATTR(FNATTR)		\
	FNATTR.setWritable(false);	\
	FNATTR.setReadable(true);	\
	FNATTR.setStorable(false);	\
	FNATTR.setKeyable(false);	\



PvDebug::PvDebug(){}

PvDebug::~PvDebug(){}


MObject PvDebug::inStart;
MObject PvDebug::inMid;
MObject PvDebug::inEnd;
MObject PvDebug::inDistMult;
MObject PvDebug::outPos;
MObject PvDebug::outRot;
MObject PvDebug::outRotX;
MObject PvDebug::outRotY;
MObject PvDebug::outRotZ;
MObject PvDebug::outResult;



void* PvDebug::creator() {
	return new PvDebug();
}


MStatus PvDebug::initialize() {
	MStatus status = MStatus::kFailure;

	MFnMatrixAttribute mAttr;
	MFnNumericAttribute nAttr;
	MFnUnitAttribute uAttr;

	inStart = mAttr.create("startMtx", "smtx", MFnMatrixAttribute::kFloat, &status);
	INPUT_ATTR(mAttr);

	inMid = mAttr.create("midMtx", "mmtx", MFnMatrixAttribute::kFloat, &status);
	INPUT_ATTR(mAttr);

	inEnd = mAttr.create("endMtx", "emtx", MFnMatrixAttribute::kFloat, &status);
	INPUT_ATTR(mAttr);

	inDistMult = nAttr.create("distanceMultiplier", "dist", MFnNumericData::kFloat, 1.0f, &status);
	nAttr.setMin(0.00001f);
	nAttr.setSoftMax(4.0f);
	nAttr.setMax(10.0f);
	INPUT_ATTR(nAttr);

	outPos = nAttr.createPoint("outPos", "op", &status);
	OUTPUT_ATTR(nAttr);

	outRotX = uAttr.create("outRotX", "orox", MFnUnitAttribute::kAngle, 0.0, &status);
	outRotY = uAttr.create("outRotY", "oroy", MFnUnitAttribute::kAngle, 0.0, &status);
	outRotZ = uAttr.create("outRotZ", "oroz", MFnUnitAttribute::kAngle, 0.0, &status);
	outRot = nAttr.create("outRot", "oro", outRotX, outRotY, outRotZ);
	OUTPUT_ATTR(nAttr);

	outResult = nAttr.createColor("result", "res", &status);
	nAttr.setUsedAsColor(true);
	OUTPUT_ATTR(nAttr);

	addAttribute(inStart);
	addAttribute(inMid);
	addAttribute(inEnd);
	addAttribute(inDistMult);
	addAttribute(outPos);
	addAttribute(outRot);
	addAttribute(outResult);

	attributeAffects(inStart, outPos);
	attributeAffects(inMid, outPos);
	attributeAffects(inEnd, outPos);
	attributeAffects(inDistMult, outPos);
	attributeAffects(inStart, outRot);
	attributeAffects(inMid, outRot);
	attributeAffects(inEnd, outRot);
	attributeAffects(inDistMult, outRot);
	attributeAffects(inStart, outResult);
	attributeAffects(inMid, outResult);
	attributeAffects(inEnd, outResult);

	return status;
}


LocationData PvDebug::retriveLocation(MFloatVector& startPos, MFloatVector& midPos, MFloatVector& endPos, float dist) {
	MFloatVector vStartEnd = endPos - startPos;
	MFloatVector vStartMid = midPos - startPos;
	float dot = vStartMid * vStartEnd;
	float proj = dot / vStartEnd.length();
	MFloatVector nStartEnd = vStartEnd.normal();
	MFloatVector vProj = nStartEnd * proj;
	MFloatVector vArrow = vStartMid - vProj;
	vArrow *= dist;
	MFloatVector vFinal = vArrow + midPos;
	MFloatVector vCross1 = vStartEnd ^ vArrow;
	vCross1.normalize();
	MFloatVector vCross2 = vCross1 ^ vArrow;
	vCross2.normalize();
	vArrow.normalize();
	double matrix[4][4] = { vArrow.x, vArrow.y, vArrow.z, 0.0f,
							vCross1.x, vCross1.y, vCross1.z, 0.0f,
							vCross2.x, vCross2.y, vCross2.z, 0.0f,
							0.0f, 0.0f, 0.0f, 1.0f };
	MMatrix mMatrix(matrix);
	MTransformationMatrix mtxFn(mMatrix);
	MEulerRotation eRot = mtxFn.eulerRotation();

	return { vFinal, eRot };
}


MFloatVector PvDebug::pvLocationRatioColor(MFloatVector& startPos, MFloatVector& midPos, MFloatVector& pvPos) {
	MFloatVector vStartMidPos = midPos - startPos;
	float startMidDist = vStartMidPos.length();
	MFloatVector vMidPvPos = pvPos - midPos;
	float midPvDist = vMidPvPos.length();

	float okRatio = startMidDist / 4.0f;
	float badRatio = 0.0f;
	float okHColor = 240.0f;
	float badHColor = 0.0f;
	float colorH = PvDebug::fit(midPvDist, badRatio, okRatio, badHColor, okHColor);
	float colorHRemaped = PvDebug::fit(colorH, 0.0f, 360.0f, 0.0f, 1.0f);
	float colorS = 1.0f;
	float colorV = 1.0f;
	MFloatVector rgb = PvDebug::convertHSVtoRGB(colorHRemaped, colorS, colorV);

	return rgb;
}


MFloatVector PvDebug::convertHSVtoRGB(float hValue, float sValue, float vValue) {
	float c = vValue * sValue; //Chroma
	float x = c * (1.0f - std::fabsf(std::fmodf(hValue / 60.0f, 2.0f) - 1.0f));
	float m = vValue - c;
	MFloatVector rgb;
	if (0.0f <= hValue && hValue < 60.0f)
		rgb = MFloatVector(c, x, 0.0f);
	else if (60.0f <= hValue && hValue < 120.0f)
		rgb = MFloatVector(x, c, 0.0f);
	else if (120.0f <= hValue && hValue < 180.0f)
		rgb = MFloatVector(0.0f, c, x);
	else if (180.0f <= hValue && hValue < 240.0f)
		rgb = MFloatVector(0.0f, x, c);
	else if (240.0f <= hValue && hValue < 300.0f)
		rgb = MFloatVector(x, 0.0f, c);
	else if (300.0f <= hValue && hValue < 360.0f)
		rgb = MFloatVector(c, 0.0f, x);
	else
		rgb = MFloatVector(0.0f, 0.0f, 0.0f);

	MFloatVector finalColor(rgb.x + m, rgb.y + m, rgb.z + m);
	
	return finalColor;
}


float PvDebug::fit(float value, float oldMin, float oldMax, float newMin, float newMax) {
	float oldRange = oldMax - oldMin;
	float newRange = newMax - newMin;
	float newValue = (((value - oldMin) * newRange) / oldRange) + newMin;
	if (value >= oldMax)
		newValue = newMax;
	else if (value <= oldMin)
		newValue = newMin;

	return newValue;
}


MStatus PvDebug::compute(const MPlug & plug, MDataBlock & dataBlock) {
	MFloatMatrix mStart = dataBlock.inputValue(inStart).asFloatMatrix();
	MFloatMatrix mMid = dataBlock.inputValue(inMid).asFloatMatrix();
	MFloatMatrix mEnd = dataBlock.inputValue(inEnd).asFloatMatrix();
	float distMult = dataBlock.inputValue(inDistMult).asFloat();

	MFloatVector vStart = mStart[3];
	MFloatVector vMid = mMid[3];
	MFloatVector vEnd = mEnd[3];

	LocationData data = PvDebug::retriveLocation(vStart, vMid, vEnd, distMult);

	if (plug == outPos) {
		MDataHandle outPosHdle = dataBlock.outputValue(outPos);
		outPosHdle.set3Float(data.translate.x, data.translate.y, data.translate.z);
		outPosHdle.setClean();
	}

	else if (plug == outRot) {
		MDataHandle outRotHdle = dataBlock.outputValue(outRot);
		MDataHandle outRotXHdle = outRotHdle.child(outRotX);
		MDataHandle outRotYHdle = outRotHdle.child(outRotY);
		MDataHandle outRotZHdle = outRotHdle.child(outRotZ);
		outRotXHdle.setMAngle(MAngle(data.rotate.x));
		outRotYHdle.setMAngle(MAngle(data.rotate.y));
		outRotZHdle.setMAngle(MAngle(data.rotate.z));
		outRotHdle.setClean();
	}

	else if (plug == outResult) {
		MFloatVector color = PvDebug::pvLocationRatioColor(vStart, vMid, data.translate);
		MDataHandle outResultHdle = dataBlock.outputValue(outResult);
		outResultHdle.setMFloatVector(color);
		outResultHdle.setClean();
	}

	else {
		return MStatus::kUnknownParameter;
	}

	return MStatus::kSuccess;
}

File: core/plugin/__DEV/Python/n_gfDebugPv.py
# -*- coding: utf-8 -*-
"""
Copyright (c) 2019 Giuliano França

Redistribution:
    Something here.

Maya Node:
    [This is a prototype version of the gfDebugPv node. You should be using the related C++ version.]
    This node is a test node and only performs test operations with one input value.

Requirements:
    Maya 2017 or above.

Todo:
    * NDA

This code supports Pylint. Rc file in project.
"""
# pylint: disable=E0401
# E0401 = Supress Maya modules import error

import math
import colorsys
import maya.api.OpenMaya as om2


def maya_useNewAPI():
    """ Function to Maya recognize the use of the Python API 2.0. """
    # pylint: disable=C0103, w0107
    pass


def INPUT_ATTR(FNATTR):
    """ Configure a input attribute. """
    # pylint: disable=C0103
    FNATTR.writable = True
    FNATTR.readable = True
    FNATTR.storable = True
    FNATTR.keyable = True


def OUTPUT_ATTR(FNATTR):
    """ Configure a output attribute. """
    # pylint: disable=C0103
    FNATTR.writable = False
    FNATTR.readable = True
    FNATTR.storable = False
    FNATTR.keyable = False


class PvDebug(om2.MPxNode):
    """ Main class of gfDebugPv node. """

    kNODE_NAME = ""
    kNODE_CLASSIFY = ""
    kNODE_ID = ""

    inStartPos = om2.MObject()
    inMidPos = om2.MObject()
    inEndPos = om2.MObject()
    outPos = om2.MObject()
    outRot = om2.MObject()
    outRotX = om2.MObject()
    outRotY = om2.MObject()
    outRotZ = om2.MObject()
    result = om2.MObject()

    def __init__(self):
        """ Constructor. """
        om2.MPxNode.__init__(self)

    @staticmethod
    def creator():
        """ Maya creator function. """
        return PvDebug()

    @staticmethod
    def initialize():
        """
        Defines the set of attributes for this node. The attributes declared in this function are assigned
        as static members to IKVChain class. Instances of IKVChain will use these attributes to create plugs
        for use in the compute() method.
        """
        startPosAttr = om2.MFnNumericAttribute()
        PvDebug.inStartPos = startPosAttr.createPoint("startPos", "sp")
        startPosAttr.writable = True
        startPosAttr.storable = True
        startPosAttr.keyable = True

        midPosAttr = om2.MFnNumericAttribute()
        PvDebug.inMidPos = midPosAttr.createPoint("midPos", "mp")
        midPosAttr.writable = True
        midPosAttr.storable = True
        midPosAttr.keyable = True

        endPosAttr = om2.MFnNumericAttribute()
        PvDebug.inEndPos = endPosAttr.createPoint("endPos", "ep")
        endPosAttr.writable = True
        endPosAttr.storable = True
        endPosAttr.keyable = True

        # Create outputs
        outPosAttr = om2.MFnNumericAttribute()
        PvDebug.outPos = outPosAttr.createPoint("outPos", "op")
        outPosAttr.writable = False
        outPosAttr.storable = True

        outRotAttr = om2.MFnUnitAttribute()
        PvDebug.outRotX = outRotAttr.create("outRotX", "orox", om2.MFnUnitAttribute.kAngle, 0.0)
        PvDebug.outRotY = outRotAttr.create("outRotY", "oroy", om2.MFnUnitAttribute.kAngle, 0.0)
        PvDebug.outRotZ = outRotAttr.create("outRotZ", "oroz", om2.MFnUnitAttribute.kAngle, 0.0)
        outRotAttr = om2.MFnNumericAttribute()
        PvDebug.outRot = outRotAttr.create("outRot", "oro", PvDebug.outRotX, PvDebug.outRotY, PvDebug.outRotZ)
        outRotAttr.writable = False
        outRotAttr.storable = True

        resultAttr = om2.MFnNumericAttribute()
        PvDebug.result = resultAttr.createColor("result", "or")
        resultAttr.writable = False
        resultAttr.storable = True
        resultAttr.usedAsColor = True

        # Add attributes
        PvDebug.addAttribute(PvDebug.inStartPos)
        PvDebug.addAttribute(PvDebug.inMidPos)
        PvDebug.addAttribute(PvDebug.inEndPos)
        PvDebug.addAttribute(PvDebug.outPos)
        PvDebug.addAttribute(PvDebug.outRot)
        PvDebug.addAttribute(PvDebug.result)
        PvDebug.attributeAffects(PvDebug.inStartPos, PvDebug.outPos)
        PvDebug.attributeAffects(PvDebug.inMidPos, PvDebug.outPos)
        PvDebug.attributeAffects(PvDebug.inEndPos, PvDebug.outPos)
        PvDebug.attributeAffects(PvDebug.inStartPos, PvDebug.outRot)
        PvDebug.attributeAffects(PvDebug.inMidPos, PvDebug.outRot)
        PvDebug.attributeAffects(PvDebug.inEndPos, PvDebug.outRot)
        PvDebug.attributeAffects(PvDebug.inStartPos, PvDebug.result)
        PvDebug.attributeAffects(PvDebug.inMidPos, PvDebug.result)
        PvDebug.attributeAffects(PvDebug.inEndPos, PvDebug.result)

    def compute(self, plug, dataBlock):
        """
        Node computation method:
            * plug is a connection point related to one of our node attributes (either an input or an output).
            * dataBlock contains the data on which we will base our computations.
        """
        # pylint: disable=R0201
        startPos = dataBlock.inputValue(PvDebug.inStartPos).asFloat3()
        midPos = dataBlock.inputValue(PvDebug.inMidPos).asFloat3()
        endPos = dataBlock.inputValue(PvDebug.inEndPos).asFloat3()

        resTrans, resRot = PvDebug.retrieveLocation(startPos, midPos, endPos)

        if plug == PvDebug.outPos:
            outPosHandle = dataBlock.outputValue(PvDebug.outPos).set3Float(resTrans[0], resTrans[1], resTrans[2])
            outPosHandle.setClean()

        elif plug == PvDebug.outRot:
            outRotHandle = dataBlock.outputValue(PvDebug.outRot)
            outRotXHandle = outRotHandle.child(PvDebug.outRotX)
            outRotYHandle = outRotHandle.child(PvDebug.outRotY)
            outRotZHandle = outRotHandle.child(PvDebug.outRotZ)
            outRotXHandle.setMAngle(om2.MAngle(resRot.x))
            outRotYHandle.setMAngle(om2.MAngle(resRot.y))
            outRotZHandle.setMAngle(om2.MAngle(resRot.z))
            outRotXHandle.setClean()
            outRotYHandle.setClean()
            outRotZHandle.setClean()
            outRotHandle.setClean()

        elif plug == PvDebug.result:
            color = PvDebug.pvLocationRatioColor(startPos, midPos, resTrans)

            outColorHandle = dataBlock.outputValue(PvDebug.result)
            outColorHandle.setMFloatVector(color)
            outColorHandle.setClean()

    @staticmethod
    def retrieveLocation(start, mid, end):
        """ Retrieve exact location of the pv. """
        vStart = om2.MVector(start[0], start[1], start[2])
        vMid = om2.MVector(mid[0], mid[1], mid[2])
        vEnd = om2.MVector(end[0], end[1], end[2])
        vStartEnd = vEnd - vStart
        vStartMid = vMid - vStart
        dotP = vStartMid * vStartEnd
        proj = float(dotP) / float(vStartEnd.length())
        nStartEnd = vStartEnd.normal()
        vProj = nStartEnd * proj
        vArrow = vStartMid - vProj
        vArrow *= 1  # Multiplier
        vFinal = vArrow + vMid
        vCross1 = vStartEnd ^ vStartMid
        vCross1.normalize()
        vCross2 = vCross1 ^ vArrow
        vCross2.normalize()
        vArrow.normalize()
        matrix = [vArrow.x, vArrow.y, vArrow.z, 0.0,
                  vCross1.x, vCross1.y, vCross1.z, 0.0,
                  vCross2.x, vCross2.y, vCross2.z, 0.0,
                  0.0, 0.0, 0.0, 1.0]
        mMatrix = om2.MMatrix(matrix)
        matrixFn = om2.MTransformationMatrix(mMatrix)
        eRot = matrixFn.rotation(asQuaternion=False)
        resTranslate = [vFinal.x, vFinal.y, vFinal.z]
        return resTranslate, eRot

    @staticmethod
    def pvLocationRatioColor(startPos, midPos, pvPos):
        """
        Get the distances of startPos-midPos and midPos-pvPos, calculate a ratio between
        these distance to get a HSV color.
        Ratio math: midPos-pvPos >= (startPos-midPos / 4)
        """
        startToMidPosDist = math.sqrt(math.pow(startPos[0] - midPos[0], 2) +
                                      math.pow(startPos[1] - midPos[1], 2) +
                                      math.pow(startPos[2] - midPos[2], 2))
        midToPvPosDist = math.sqrt(math.pow(midPos[0] - pvPos[0], 2) +
                                   math.pow(midPos[1] - pvPos[1], 2) +
                                   math.pow(midPos[2] - pvPos[2], 2))
        okRatio = startToMidPosDist / 4
        badRatio = 0.0
        okHColor = 180.0
        badHColor = 0.0
        colorH = PvDebug.fit(midToPvPosDist, badRatio, okRatio, badHColor, okHColor)
        colorH = PvDebug.fit(colorH, 0.0, 360.0, 0.0, 1.0)
        colorS = 1.0
        colorV = 1.0
        rgb = colorsys.hsv_to_rgb(colorH, colorS, colorV)
        color = om2.MFloatVector(rgb[0], rgb[1], rgb[2])
        return color

    @staticmethod
    def fit(value, oldMin, oldMax, newMin, newMax):
        """
        Fit function: (((value-oldMin)*(newMax-newMin))/(oldMax-oldMin))+newMin
        """
        oldRange = oldMax - oldMin
        newRange = newMax - newMin
        newValue = (((value - oldMin) * newRange) / oldRange) + newMin
        if value >= oldMax:
            newValue = newMax
        elif value <= oldMin:
            newValue = newMin
        return newValue
